ppafm/ocl/AFMulator.py

```diff
@@ -392,6 +392,87 @@ def evalAFM(self, X=None):
+        return afmulator
+
+    def load_params(self, file_path='./params.ini'):
+        '''
+        Update the parameters of this AFMulator instance from a params.ini file.
+
+        Arguments:
+            file_path: str. Path to the params.ini file to load.
+        '''
+        params, sample_lvec = _get_params(file_path)
+        if params['tipStiffness'] is not None:
+            self.scanner.stiffness = np.array(params['tipStiffness'], dtype=np.float32) / -PPU.eVA_Nm
+        self.iZPP = params['iZPP']
+        self.tipR0 = params['tipR0']
+        self.f0Cantilever = params['f0Cantilever']
+        self.kCantilever = params['kCantilever']
+        self.npbc = params['npbc']
+        self.A_pauli = params['A_pauli']
+        self.setScanWindow(params['scan_window'], params['scan_dim'], params['df_steps'])
+        self.setLvec(params['lvec'], params['pixPerAngstrome'])
+        self.setRho(params['rho'], params['sigma'], params['B_pauli'])
+        self.sample_lvec = sample_lvec
+
+    def save_params(self, file_path='./params.ini'):
+        '''
+        Save the parameters of this AFMulator instance to a params.ini file.
+
+        Arguments:
+            file_path: str. Path to the file where parameters are saved.
+        '''
+        k = self.scanner.stiffness * -PPU.eVA_Nm
+        nDim = self.forcefield.nDim
+        scan_step = (
+            (self.scan_window[1][0] - self.scan_window[0][0]) / (self.scan_dim[0] - 1),
+            (self.scan_window[1][1] - self.scan_window[0][1]) / (self.scan_dim[1] - 1),
+            (self.scan_window[1][2] - self.scan_window[0][2]) / self.scan_dim[2]
+        )
+        with open(file_path, 'w') as f:
+            f.write(f'probeType {self.iZPP}\n')
+            if isinstance(self._rho, dict):
+                if len(self._rho) > 1:
+                    warnings.warn('More than one tip multipole type specified. Only writing the first one into params.ini.')
+                tip = list(self._rho)[0]
+                f.write(f'tip {tip}\n')
+                f.write(f'charge {self._rho[tip]}\n')
+                f.write(f'sigma {self.sigma}\n')
+            f.write(f'stiffness {k[0]} {k[1]} {k[3]}\n')
+            f.write(f'r0Probe {self.tipR0[0]} {self.tipR0[1]} {self.tipR0[2]}\n')
+            f.write(f'PBC {(np.array(self.npbc) > 0).any()}\n')
+            f.write(f'nPBC {self.npbc[0]} {self.npbc[1]} {self.npbc[2]}\n')
+            if self.sample_lvec is not None:
+                f.write(f'gridA {self.sample_lvec[0, 0]} {self.sample_lvec[0, 1]} {self.sample_lvec[0, 2]}\n')
+                f.write(f'gridB {self.sample_lvec[1, 0]} {self.sample_lvec[1, 1]} {self.sample_lvec[1, 2]}\n')
+                f.write(f'gridC {self.sample_lvec[2, 0]} {self.sample_lvec[2, 1]} {self.sample_lvec[2, 2]}\n')
+            f.write(f'FFgrid0 {self.lvec[0, 0]} {self.lvec[0, 1]} {self.lvec[0, 2]}\n')
+            f.write(f'FFgridA {self.lvec[1, 0]} {self.lvec[1, 1]} {self.lvec[1, 2]}\n')
```

Comment:
Wow, I actually did not know it is possible to `sprintf` like this in python :)

Reply:
They are called f-strings. They were introduced in Python 3.6. It looks pretty nice and supposedly also executes faster than the old string formatting style: https://realpython.com/python-f-strings/.